lib/Providers/OpenIDConnectProvider.php

```diff
@@ -0,0 +1,186 @@
+
+    /**
+     * @var string
+     */
+    protected $issuer;
+
+    /**
+     * @var array
+     */
+    private $openIdConfiguration;
+
+    /**
+     * @var string
+     */
+    private $responseError = 'error';
+
+    public function __construct(array $options = [], array $collaborators = [])
+    {
+        parent::__construct($options, $collaborators);
+
+        if (!array_key_exists('issuer', $options))
+        {
+            throw new \InvalidArgumentException(
+                'Required options not defined: issuer'
+            );
+        }
+        $this->issuer = $options['issuer'];
+    }
+
+    protected function getScopeSeparator()
+    {
+        return ' ';
+    }
+
+    protected function getDefaultScopes()
+    {
+        return 'openid profile';
+    }
+
+    protected function checkResponse(ResponseInterface $response, $data)
+    {
+        $error = null;
+        if (!empty($data[$this->responseError])) {
+            $error = $data[$this->responseError];
+            if (!is_string($error)) {
+                $error = var_export($error, true);
+            }
+        }
+        if ($error || $response->getStatusCode() >= 400) {
+            throw new IdentityProviderException($error, 0, $data);
+        }
+    }
+
+    protected function createResourceOwner(array $response, AccessToken $token)
+    {
+        return new GenericResourceOwner($response, 'id');
+    }
+
+    /**
+     * Do any required verification of the id token and return an array of decoded claims
+     *
+     * @param string $id_token Raw id token as string
+     * @return array associative array of claims decoded from the id token
+     */
+    public function verifyIdToken($id_token)
+    {
+        try {
+            $keys = $this->getSigningKeys();
+            $claims = JWT\JWT::decode($id_token, $keys, ['RS256']);
+            if ($claims->aud !== $this->clientId) {
+                throw new IdentityProviderException("ID token has incorrect audience", 0, $claims->aud);
+            }
+            if ($claims->iss !== $this->issuer) {
+                throw new IdentityProviderException("ID token has incorrect issuer", 0, $claims->iss);
+            }
+        } catch (\UnexpectedValueException $e) {
+            throw new IdentityProviderException("ID token validation failed", 0, $e->getMessage());
+        }
+    }
+
+    protected function prepareAccessTokenResponse(array $result)
+    {
+        $result = parent::prepareAccessTokenResponse($result);
+        $this->verifyIdToken($result['id_token']);
+        return $result;
+    }
+
+    protected function getOpenIDConfiguration() {
+        if (isset($this->openIdConfiguration)) {
+            return $this->openIdConfiguration;
+        }
+
+        $config = $this->getParsedResponse($this->getRequest('GET', $this->issuer . self::CONFIGURATION_PATH));
+        $this->openIdConfiguration = $config;
+        return $config;
+    }
+
+    public static function base64url_decode($input) {
```

Comment:
Not sure this makes sense as a public method. It doesn't feel like something this class should offer as an interface. If it's only intended for internal use, I'd suggest making it private or at least protected.

Reply:
fair point.